README
This is the 2.3 (USENET) version of the Elm Mail System

See the NOTICE and Instruct files for further details.

It is IMPERITIVE that all users rerun newalias after installing
Elm 2.3 when upgrading from a previous version.  Elm's behavoir
with aliases could be unpredictible if this step is not performed.

Where to find more info on Elm:
   Much discussion on Elm including interim bug fixes, work arounds
and future plans occurs in the Usenet news group comp.mail.elm.  Also
a monthly status report on Elm is posted there.  This status report
lists the archive sites that have the patches to Elm as well as the
lastest version.

Patches to Elm are posted to comp.mail.elm and comp.sources.bugs as
soon as they are released.  They are posted to comp.sources.unix shortly
thereafter, to allow time for feedback of problems in the patches.
Patches should be available from the archive sites, or from the
archive server.  Mail archive-server@DSI.COM for details on how to
use the archive server program.  Ask it for help.

Limitations/Problems you might encounter in compiling and installing Elm:

	On some Unix 5.3.2 systems, if only a runtime version of the
O/S has been installed, not all the include files exist for compiling
curses.c.  The ptem.h include file contains the window sizing structure
on this version.  Either comment out the window sizing code, or install
the remaining include files from the development system.

	If you run a nonstandard configuration of Mail, such as Xenix
running smail, Configure can get confused as to where to place items.
Be sure and check the config.sh file for the correct placement before
continuning.  If changes are necessary, rerun Configure and fix the
file before exiting.

	On SCO Xenix, if you are all mail is from user anonymous,
this is because the mail delivery agent should be
/usr/lib/mail/execmail instead of /usr/bin/rmail or /bin/rmail.

	On Next's it is necessary to answer -bsd for the
Any additional cc flags? question and -lsys_s for the
Any additional libraries? question within Configure.

	On some systems, especially those based on the AT&T Port to
286's the -O flag of the compiler produces improper code causing
segmentation violations.  If this happens, recompile the code without
the -O flag.  This has been seen with Microport SysV/AT type systems.

	The Configuration script has been known to exceed the default
stack size in Unix 286 sh's.  If Configure does not run correctly on
this type of machine increase the stack size and rerun it.  On
Microport SysV/AT machines, Configure might run correctly under ksh.
Obtain ksh from Microport (available to current version owners without
additional charge) and rerun Configure using it.  Xenix 286 users may
have the same problem, but they can work around it by changing the
stack size within the shell as reported a Xenix 286 Elm user:
> I had the same problems as everyone else is reporting on Microports *nix under
> SCO XENIX 2.2.1.
>
> The solution is simple, up the stack size for /bin/sh. I used
>
> # mv /bin/sh /bin/sh.old
> # cp /bin/sh.old /bin/sh
> # fixhdr -F 8000 /bin/sh
>
> This may seem a bit over the top, but I put it back after!
>
> # mv /bin/sh /bin/sh.rm
> # mv /bin/sh.old /bin/sh
>
> Then wait till nobody is using /bin/sh.rm then
>
> # rm /bin/sh.rm
>
> That way you preserve your old shell, ( ie you don't break it ), but you
> get to use Configure without bus errors etc.
>
> I hope that of use
>
> Keith
> --
> UUCP ..!uunet!mcvax!ukc!slxsys!g4lzv!keith  | Keith Brazington
> Smart mail  keith@g4lzv.co.uk		    | 5b Northgate Rochester Kent UK
> Ampanet  [44.131.8.1] and [44.131.8.3]	    | +44 634 811594 Voice
> Packet  G4LZV @ GB7UWS -- G4LZV USENET BB --| +44 634 401210 Data v22,v22bis
You might have to experiment and try values from 7000 to 8800.

	Also on Microport SysV/AT Machines, the C compiler produces
improper code for one of the arithmetic calls.  To fix this problem it
is necessary to reduce the complexity of the statement, as reported by
one of our testers here is the symptom and his patch.  Being this is a
compiler bug on only one system, we make the information available, but
not incorporate it in the main release.  Note, this may effect other
areas of Elm, and in the future, Microport may even fix this problem.
> I finally tracked down the bug that was causing the lengthy delays when
> the first message was displayed.  The Microport 80286 C compiler was
> generating bad code for the computation of padding in showmsg.c.  The
> compiler generated scratch variables in the expression were being
> located at weird offsets in the stack segment.  This caused the program
> to stall while the kernel attempted to grow the stack segment to a size
> that was large enough to contain the scratch variables.  This explains
> why it only happened the first time a message was displayed.  Here is a
> patch that fixes the problem:
>
> *** showmsg.c.dist	Fri Mar 17 21:08:37 1989
> --- showmsg.c	Sat Mar 18 06:14:04 1989
> ***************
> *** 280,289
>   	           atoi(current_header->year), current_header->time);
>
>   	  /* truncate or pad title2 portion on the right
> ! 	   * so that line fits exactly */
> ! 	  padding =
> ! 	    COLUMNS -
> ! 	    (strlen(title1) + (buf_len=strlen(title2)) + strlen(title3));
>
>   	  sprintf(titlebuf, "%s%-*.*s%s\n", title1, buf_len+padding,
>   	      buf_len+padding, title2, title3);
>
> --- 280,292 -----
>   	           atoi(current_header->year), current_header->time);
>
>   	  /* truncate or pad title2 portion on the right
> ! 	   * so that line fits exactly, expression has been
> ! 	   * simplified to avoid bug in Microport 80286
> ! 	   * C compiler */
> ! 	  padding = COLUMNS;
> ! 	  padding -= strlen(title1);
> ! 	  padding -= (buf_len = strlen(title2));
> ! 	  padding -= strlen(title3);
>
>   	  sprintf(titlebuf, "%s%-*.*s%s\n", title1, buf_len+padding,
>   	      buf_len+padding, title2, title3);
>
> --
> John A. Limpert
> UUCP:	johnl@n3dmc.UUCP, johnl@n3dmc.UU.NET, uunet!n3dmc!johnl


	Some versions of GNUmake are slightly incompatible with
standard make.  If you plan on using GNUmake instead of standard make,
you might need to change instances of:
	-$(MAKEFLAGS)
to:
	$(MAKEFLAGS)
in all Makefiles that contain that construct.

>From pyrdc!uunet!mcvax!sauna.hut.fi!jkp Thu Jul 20 23:06:10 1989
>From: Jyrki Kuoppala <pyrdc!uunet!mcvax!cs.hut.fi!jkp>
>Subject: ELM diff for Altos Worknet
>Organization: Helsinki University of Technology, Finland.
>There's a bug in the Altos Worknet that makes it impossible to change
>a file's access time if the file is on another system (via Worknet).
>To make elm work if you use /usr/mail over Worknet, you need to apply
>the following patch and define WORKNET when compiling leavembox.c.
>
>Note: WORKNET needs to be defined somewhere, the easiest thing is
>to add it manually to src/Makefile.  That's how I did it.
>
>*** elm2.2/src/leavembox.c	Tue Jul 18 20:50:32 1989
>--- elm2.2.altos/src/leavembox.c	Tue Jul 18 23:36:11 1989
>***************
>*** 557,562 ****
>--- 557,565 ----
>  	utime_buffer.modtime= buf.st_mtime;
>  #endif
>
>+ #ifndef WORKNET
>+ 	/* Braindamaged WorkNet from Altos doesn't know how to change
>+ 	   a file's access time if it is on another system */
>  #ifdef BSD
>  	if (utime(cur_folder, utime_buffer) != 0) {
>  #else
>***************
>*** 569,574 ****
>--- 572,578 ----
>  	  error2("Error %s trying to change file %s access time.",
>  		   error_name(errno), cur_folder);
>  	}
>+ #endif  /* WORKNET */
>
>
>  	mailfile_size = bytes(cur_folder);
>
>
>Jyrki Kuoppala    Helsinki University of Technology, Finland.
>Internet :        jkp@cs.hut.fi           [128.214.3.119]
>BITNET :          jkp@fingate.bitnet      Gravity is a myth, the Earth sucks!

					Syd Weinstein
					Elm Coordinator
					elm@DSI.COM
					(dsinc!elm)


Changes
This file describes major changes made to the Elm source distribution
between Releases 2.2 and 2.3.  Please remember that Elm is a volunteer
project, we did what we had time for, and perhaps, what interested us.
If you have a suggestion, or a fix, submit it, and better yet,
volunteer to impliment it.

        Syd Weinstein		Sat Apr 28 22:16:46 EDT 1990

Complete rewrite of date parsing to allow for RFC822 dates and time zones.

DEL is no longer a syn. for x.

Add BSD style screen sizing when the IOCTL calls exist.

Remove direct access to the L.sys/Systems file and only use uuname.

Add -v option for Elm version and options information.

Allows specifying the temp directory to be used by Elm.

Allow for restricting shell escapes from Elm.  (But of course, not
from any pager or editor Elm would invoke)

Add PID type locking to the lock file.

Make Elm confirm an x command if changes pending, allow X to be the
old non confirming x command.

Allow for disabling the X-Mailer header if desired.

Complete revamp of Configure using new Metaconfig modules.

Add some checking for space problems and force an emergency exit in those
cases.

Allow the alias file to be used across heterogenous networks by
keeping it in network byte order if NFS is available.

Add support for MMDF style mailboxes.

Aliases are now case insensitive as per survey results

Recursive aliases now work again.

And many bug fixes.

************************************************************************


src/init.c

static char rcsid[] = "@(#)$Id: init.c,v 4.1.1.4 90/12/05 14:34:08 syd Exp $";

/*******************************************************************************
 *  The Elm Mail System  -  $Revision: 4.1.1.4 $   $State: Exp $
 *
 * 			Copyright (c) 1986, 1987 Dave Taylor
 * 			Copyright (c) 1988, 1989, 1990 USENET Community Trust
 *******************************************************************************
 * Bug reports, patches, comments, suggestions should be sent to:
 *
 *	Syd Weinstein, Elm Coordinator
 *	elm@DSI.COM			dsinc!elm
 *
 *******************************************************************************
 * $Log:	init.c,v $
 * Revision 4.1.1.4  90/12/05  14:34:08  syd
 * fix dropping of unused vars, dropped to much
 * From: Syd
 *
 * Revision 4.1.1.3  90/10/24  15:33:11  syd
 * Remove variables no longer used
 * From: W. David Higgins
 *
 * Revision 4.1.1.2  90/08/02  21:57:56  syd
 * The newly introduced function 'stricmp' has a name conflict with a libc
 * function under SunOS 4.1.  Changed name to istrcmp.
 * From: scs@lokkur.dexter.mi.us (Steve Simmons)
 *
 * Revision 4.1.1.1  90/07/12  23:19:17  syd
 * Make domain name checking case independent
 * From: Syd, reported by Steven Baur
 *
 * Revision 4.1  90/04/28  22:43:15  syd
 * checkin of Elm 2.3 as of Release PL0
 *
 *
 ******************************************************************************/

/***** Initialize - read in all the defaults etc etc
*****/

#include "headers.h"
#include "patchlevel.h"

#ifdef TERMIOS
#  include <termios.h>
   typedef struct termios term_buff;
#else
# ifdef TERMIO
#  include <termio.h>
#  define tcgetattr(fd,buf)	ioctl((fd),TCGETA,(buf))
   typedef struct termio term_buff;
# else
#  include <sgtty.h>
#  define tcgetattr(fd,buf)	ioctl((fd),TIOCGETP,(buf))
   typedef struct sgttyb term_buff;
# endif
#endif

#ifdef PWDINSYS
#  include <sys/pwd.h>
#else
#  include <pwd.h>
#endif

#ifdef I_TIME
#  include <time.h>
#endif
#ifdef I_SYSTIME
#  include <sys/time.h>
#endif

#include <signal.h>
#include <ctype.h>
#include <errno.h>

#ifdef BSD
#undef toupper
#undef tolower
#endif

#ifndef OS2
extern int errno;		/* system error number on failure */
#endif
extern char version_buff[];

char *error_name(), *error_description();

char *getenv(), *getlogin(), *strcpy(), *strcat(), *index();
unsigned short getgid(), getuid();
void exit();
#ifndef	_POSIX_SOURCE
struct passwd *getpwuid();
#endif
char *get_full_name();

#ifdef VOIDSIG
  void
#else
  int
#endif
#ifdef SIGTSTP
	sig_user_stop(), sig_return_from_user_stop(),
#endif
	quit_signal(), term_signal(), ill_signal(),
	fpe_signal(),  bus_signal(),  segv_signal(),
	alarm_signal(), pipe_signal(), hup_signal();

initialize(requestedmfile)
char *requestedmfile;	/* first mail file to open, empty if the default */
{
	/** initialize the whole ball of wax.
	**/
	struct passwd *pass;
	register int i, hostlen, domlen;

#if defined(SIGVEC) & defined(SV_INTERRUPT)
	struct sigvec alarm_vec;
#endif
	char     buffer[SLEN], *cp;

	sprintf(version_buff, "%s PL%d", VERSION, PATCHLEVEL);
	Raw(ON);

	/* save original user and group ids */
	userid  = getuid();
	groupid = getgid();

	/* make all newly created files private */
	original_umask = umask(077);

	/* Get username (logname), home (login directory), and full_username
	 * (part of GCOS) field from the password entry for this user id.
	 * Full_username will get overridden by fullname in elmrc, if defined.
	 */

	if((pass = getpwuid(userid)) == NULL) {
	  error("You have no password entry!");
	  Raw(OFF);
	  exit(1);
	}
	strcpy(username, pass->pw_name);
	strcpy(home, pass->pw_dir);

	if((cp = get_full_name(username)) != NULL)
	  strcpy(full_username, cp);
	else
	  strcpy(full_username, username);	/* fall back on logname */

#ifdef DEBUG
	if (debug) {		/* setup for dprint() statements! */
	  char newfname[SLEN], filename[SLEN];

	  sprintf(filename, "%s/%s", home, DEBUGFILE);
	  if (access(filename, ACCESS_EXISTS) == 0) {	/* already one! */
	    sprintf(newfname,"%s/%s", home, OLDEBUG);
	    (void) rename(filename, newfname);
	  }

	  /* Note what we just did up there: we always save the old
	     version of the debug file as OLDEBUG, so users can mail
	     copies of bug files without trashing 'em by starting up
	     the mailer.  Dumb, subtle, but easy enough to do!
 	  */

	  if ((debugfile = fopen(filename, "w")) == NULL) {
	    debug = 0;	/* otherwise 'leave' will try to log! */
	    leave(fprintf(stderr,"Could not open file %s for debug output!\n",
		  filename));
	  }
	  chown(filename, userid, groupid); /* file owned by user */

	  fprintf(debugfile,
     "Debug output of the ELM program (at debug level %d).  Version %s\n\n",
		  debug, version_buff);
	}
#endif

	/*
	 * If debug level is fairly low, ignore keyboard signals
	 * until the screen is set up.
	 */
	if (debug < 5) {
	  signal(SIGINT,  SIG_IGN);
	  signal(SIGQUIT, SIG_IGN);
	}

	if(!check_only && !batch_only) {
	  if ((i = InitScreen()) < 0) {
	    if (i == -1) {
	      printf(
"Sorry, but you must specify what type of terminal you're on if you want to\r\n");
	      printf(
"run the \"elm\" program. (You need your environment variable \"TERM\" set.)\r\n"
		     );
	      dprint(1,(debugfile,"No $TERM variable in environment!\n"));
	    }
	    else if (i == -2) {
	      printf(
"You need a cursor-addressable terminal to run \"elm\" and I can't find any\r\n");
	      printf(
"kind of termcap entry for \"%s\" - check your \"TERM\" setting...\r\n",
		   getenv("TERM"));
	      printf(
"Or check your TERMCAP setting or termcap database file.\r\n");
	      dprint(1,
		(debugfile,"$TERM variable is an unknown terminal type!\n"));
	    } else {
	      printf("Failed trying to initialize your terminal entry: unknown return code %d\r\n", i);
	      dprint(1, (debugfile, "Initscreen returned unknown code: %d\r\n",
		  i));
	    }
	    Raw(OFF);
	    exit(1);	/* all the errors share this exit statement */
	  }
          EndBold();
          EndHalfbright();
          EndInverse();
	}

	if (debug < 5) {	/* otherwise let the system trap 'em! */
	  signal(SIGQUIT, quit_signal);		/* Quit signal 	            */
	  signal(SIGTERM, term_signal); 	/* Terminate signal         */
	  signal(SIGILL,  ill_signal);		/* Illegal instruction      */
	  signal(SIGFPE,  fpe_signal);		/* Floating point exception */
#ifndef OS2
	  signal(SIGBUS,  bus_signal);		/* Bus error  		    */
	  signal(SIGHUP,  hup_signal);		/* HangUp (line dropped)    */
#endif
	  signal(SIGSEGV, segv_signal);		/* Segmentation Violation   */
	}
	else {
	  dprint(3,(debugfile,
  "\n*** Elm-Internal Signal Handlers Disabled due to debug level %d ***\n\n",
		    debug));
	}
#if defined(SIGVEC) & defined(SV_INTERRUPT)
	alarm_vec.sv_handler = alarm_signal;
	alarm_vec.sv_flags = SV_INTERRUPT;
	sigvec (SIGALRM, &alarm_vec, (struct sigvec *)0);	/* Process Timer Alarm	    */
#else
	signal(SIGALRM, alarm_signal);		/* Process Timer Alarm      */
#endif
#ifndef OS2
	signal(SIGPIPE, pipe_signal);		/* Illegal Pipe Operation   */
#endif
#ifdef SIGTSTP
	signal(SIGTSTP, sig_user_stop);		/* Suspend signal from tty  */
	signal(SIGCONT, sig_return_from_user_stop);	/* Continue Process */
#endif

	get_term_chars();

	/*
	 * Get the host name as per configured behavior.
	 */
#ifdef HOSTCOMPILED
	strncpy(hostname, HOSTNAME, sizeof(hostname) - 1);
	hostname[sizeof(hostname) - 1] = '\0';
#else
	gethostname(hostname, sizeof(hostname));
#endif

	/*
	 * now get the domain name, used to build the full name
	 */
	gethostdomain(hostdomain, sizeof(hostdomain));

	/*
	 * now the tough part:
	 *	we need to make three variables out of this stuff:
	 *	hostname = just the hostname, as in bangpaths,
	 *		this is whatever the user gave us so far,
	 *		we wont change this one
	 *	hostdomain = this is the domain considered local to this
	 *		machine, and should be what we got above.
	 *	hostfullname = this is the full FQDN of this machine,
	 *		and is a strange combination of the first two.
	 *	if tail(hostname) == hostdomain
	 *		then hostfullname = hostname
	 *			ie: node.ld.domain.type, ld.domain.type -> node.ld.domain.type
	 *	else if hostname == hostdomain + 1
	 *		then hostfullname = hostname
	 *			ie: domain.type, .domain.type -> domain.type
	 *
	 *	else hostfullname = hostname + hostdomain
	 *			ie: host, .domain.type -> host.domain.type
	 * lost yet?
	 */
	hostlen = strlen(hostname);
	domlen = strlen(hostdomain);
	if (hostlen >= domlen) {
	  if (istrcmp(&hostname[hostlen - domlen], hostdomain) == 0)
	    strcpy(hostfullname, hostname);
	  else {
	    strcpy(hostfullname, hostname);
	    strcat(hostfullname, hostdomain);
	  }
	} else {
	  if (istrcmp(hostname, hostdomain + 1) == 0)
	    strcpy(hostfullname, hostname);
	  else {
	    strcpy(hostfullname, hostname);
	    strcat(hostfullname, hostdomain);
	  }
	}

#ifdef OS2
	getfromdomain(hostfromname, sizeof(hostfromname));
#else
	strcpy(hostfromname, hostfullname);
#endif

	/* Determine the default mail file name.
	 *
	 * First look for an environment variable MAIL, then
	 * use then mailhome if it is not found
	 */
	if ((cp = getenv("MAIL")) == NULL)
#ifdef OS2
	    if (maildir)
		sprintf(defaultfile, "%s%s/newmail%s", mailhome, username, mailext);
	    else
#endif
		sprintf(defaultfile, "%s%s%s", mailhome, username, mailext);
	else
		strcpy(defaultfile, cp);

	/* Determine options that might be set in the .elm/elmrc */
	read_rc_file();

	/* Determine the mail file to read */
	if (*requestedmfile == '\0')
	  strcpy(requestedmfile, defaultfile);
	else if(!expand_filename(requestedmfile, FALSE)) {
	    Raw(OFF);
	    exit(0);
        }
	if (check_size)
	  if(check_mailfile_size(requestedmfile) != 0) {
	      Raw(OFF);
	      exit(0);
	  }

	/* check for permissions only if not send only mode file */
	if (! mail_only) {
	  if ((errno = can_access(requestedmfile, READ_ACCESS)) != 0) {
	    if (strcmp(requestedmfile, defaultfile) != 0 || errno != ENOENT) {
	      dprint(1, (debugfile,
		    "Error: given file %s as folder - unreadable (%s)!\n",
		    requestedmfile, error_name(errno)));
	      fprintf(stderr,"Can't open folder '%s' for reading!\n",
		    requestedmfile);
	      Raw(OFF);
	      exit(1);
	    }
	  }
	}

	/** check to see if the user has defined a LINES or COLUMNS
	    value different to that in the termcap entry (for
	    windowing systems, of course!) **/

	ScreenSize(&LINES, &COLUMNS);

	if ((cp = getenv("LINES")) != NULL && isdigit(*cp)) {
	  sscanf(cp, "%d", &LINES);
	  LINES -= 1;	/* kludge for HP Window system? ... */
	}

	if ((cp = getenv("COLUMNS")) != NULL && isdigit(*cp))
	  sscanf(cp, "%d", &COLUMNS);

	/** fix the shell if needed **/

#ifndef OS2
	if (shell[0] != '/') {
	   sprintf(buffer, "/bin/%s", shell);
	   strcpy(shell, buffer);
	}
#endif

	if (! mail_only && ! check_only) {

	  /* get the cursor control keys... */

	  cursor_control = FALSE;

	  if ((cp = return_value_of("ku")) != NULL) {
	    strcpy(up, cp);
	    if ((cp = return_value_of("kd")) != NULL) {
	      strcpy(down, cp);
	      if ((cp = return_value_of("kl")) != NULL) {
		strcpy(left, cp);
		if ((cp = return_value_of("kr")) != NULL) {
		  strcpy(right, cp);
		  cursor_control = TRUE;
		  transmit_functions(ON);
		}
	      }
	    }
	  }
	  if (!arrow_cursor) 	/* try to use inverse bar instead */
	    if (return_value_of("so") != NULL && return_value_of("se") != NULL)
	        has_highlighting = TRUE;
	}

	/** clear the screen **/
	if(!check_only && !batch_only)
	  ClearScreen();

	if (! mail_only && ! check_only) {
	  if (mini_menu)
	    headers_per_page = LINES - 13;
	  else
	    headers_per_page = LINES -  8;	/* 5 more headers! */

	  newmbox(requestedmfile, FALSE);	/* read in the folder! */
	}

#ifdef DEBUG
	if (debug >= 2 && debug < 10) {
	  fprintf(debugfile,
"hostname = %-20s \tusername = %-20s \tfullname = %-20s\n",
	         hostname, username, full_username);

	  fprintf(debugfile,
"home     = %-20s \teditor   = %-20s \trecvd_mail  = %-20s\n",
		 home, editor, recvd_mail);

	  fprintf(debugfile,
"cur_folder   = %-20s \tfolders  = %-20s \tprintout = %-20s\n",
		 cur_folder, folders, printout);

	  fprintf(debugfile,
"sent_mail = %-20s \tprefix   = %-20s \tshell    = %-20s\n\n",
		sent_mail, prefixchars, shell);

	  if (local_signature[0])
	    fprintf(debugfile, "local_signature = \"%s\"\n",
			local_signature);
	  if (remote_signature[0])
	    fprintf(debugfile, "remote_signature = \"%s\"\n",
			remote_signature);
	  if (local_signature[0] || remote_signature[0])
	    fprintf(debugfile, "\n");
	}
#endif
}

get_term_chars()
{
	/** This routine sucks out the special terminal characters
	    ERASE and KILL for use in the input routine.  The meaning
            of the characters are (dare I say it?) fairly obvious... **/

	term_buff term_buffer;

	if (tcgetattr(STANDARD_INPUT,&term_buffer) == -1) {
	  dprint(1, (debugfile,
		   "Error: %s encountered on ioctl call (get_term_chars)\n",
		   error_name(errno)));
	  /* set to defaults for terminal driver */
	  backspace = BACKSPACE;
	  kill_line = ctrl('U');
	}
	else {
#if defined(TERMIO) || defined(TERMIOS)
	  backspace = term_buffer.c_cc[VERASE];
	  kill_line = term_buffer.c_cc[VKILL];
#else
	  backspace = term_buffer.sg_erase;
	  kill_line = term_buffer.sg_kill;
#endif
	}
}
